core/modules/replica/IndexJob.h

```diff
@@ -0,0 +1,292 @@
+     */
+    IndexJobResult const& getResultData() const;
+
+    std::list<std::pair<std::string,std::string>> extendedPersistentState() const final;
+    std::list<std::pair<std::string,std::string>> persistentLogData() const final;
+
+protected:
+
+    void startImpl(util::Lock const& lock) final;
+
+    void cancelImpl(util::Lock const& lock) final;
+
+    void notify(util::Lock const& lock) final;
+
+private:
+
+    IndexJob(std::string const& database,
+             bool hasTransactions,
+             uint32_t transactionId,
+             bool allWorkers,
+             Destination destination,
+             std::string const& destinationPath,
+             Controller::Ptr const& controller,
+             std::string const& parentJobId,
+             CallbackType const& onFinish,
+             Job::Options const& options);
+
+    /**
+     * The callback function to be invoked on a completion of requests
+     * targeting workers.
+     */
+    void _onRequestFinish(IndexRequest::Ptr const& request);
+
+    /**
+     * Extract data from the successfully completed requests. The completion
+     * state of the request will be evaluated by the method.
+     *
+     * @param lock
+     *   on the mutex Job::_mtx to be acquired for protecting the object's state
+     *
+     * @param request
+     *   the request to extract the data to be processed
+     */
+    void _processRequestData(util::Lock const& lock,
+                             IndexRequest::Ptr const& request);
+
+    /**
+     * Lunch a batch of requests with a total number not to exceed the specified
+     * limit.
+     *
+     * @param lock
+     *   on the mutex Job::_mtx to be acquired for protecting the object's state
+     *
+     * @param worker
+     *   the name of a worker the requests to be sent to
+     * 
+     * @param maxRequests
+     *   the maximum number of requests to be launched
+     *
+     * @return a collection of requests launched
+     */
+    std::list<IndexRequest::Ptr> _launchRequests(util::Lock const& lock,
+                                                 std::string const& worker,
+                                                 size_t maxRequests=1);
+
+    /**
+     * Roll back a database transaction should the one be still open
+     * for destination TABLE. The method won't have any effect for other
+     * scenarios.
+     *
+     * @param func  the name of a method/function to report errors
+     */
+    void _rollbackTransaction(std::string const& func);
+
+private:
+
+    // Input parameters
+
+    std::string const _database;
+    bool        const _hasTransactions;
+    uint32_t    const _transactionId;
+    bool        const _allWorkers;
+    Destination const _destination;
+    std::string const _destinationPath;
+
+    CallbackType _onFinish;     /// @note is reset when the job finishes
+
+    /// A collection of chunks to be processed at specific workers
+    std::map<std::string, std::list<unsigned int>> _chunks;
```

Comment:
It depends. In my case the list stores small elements of 4 bytes each. And the only operations I care about in this application is `push_back` and `pop_front` (and `front`). According to this article (comparing the performance of the relevant containers), the performance of all container (`list`, `deque`, `vector` and pre-sized `vector`) is about the same for small-size elements (16 bytes in the article):
https://baptiste-wicht.com/posts/2012/12/cpp-benchmark-vector-list-deque.html
And things are quite differet (`deque` is the winner) for the large-size element types (tis is not my case). And I don't care about the performance of `find`. 
I couldn't find any info on a performance of the `pop_front` in that article.

Reply:
- I don't trust random articles from random people on internet. 
- The 16-bytes example that you have chosen is where that guy uses "non-trivial" data type with complicated copy constructor that clearly dominates all operations
- If you do not care about performance always choose `vector`

Reply:
If you just care about `push_back`, `pop_front`, and `front`, why not be explicit and just use `std::queue`?  (It will default to using a `std::deque`, so Andy can also be happy.  And you'll show some mercy to the poor heap...

Reply:
Good suggestion! I will look at switching to `std::queue`.